lib/SILOptimizer/Utils/Generics.cpp

```diff
@@ -244,6 +244,470 @@ void ReabstractionInfo::createSubstitutedAndSpecializedTypes() {
+      assert(isa<GenericTypeParamType>(BaseTy));
+      return CalleeInterfaceToCallerArchetypeMap.getMap()
+          .lookup(cast<SubstitutableType>(BaseTy->getCanonicalType()))
+          ->getCanonicalType()
+          ->hasArchetype();
+    };
+
+    auto GetBaseGenericTypeParamType =
+        [](CanType Ty) -> GenericTypeParamType * {
+      auto BaseTy = Ty;
+      while (auto DepTy = dyn_cast<DependentMemberType>(BaseTy)) {
+        BaseTy = DepTy->getBase()->getCanonicalType();
+      };
+      assert(isa<GenericTypeParamType>(BaseTy));
+      return cast<GenericTypeParamType>(BaseTy);
+    };
+
+    // Create the updated SubstitutionMap to be used by the cloner.
+    for (auto Pair : CalleeInterfaceToCallerArchetypeMap.getMap()) {
+      auto Ty = Pair.first;
+      auto CanTy = Pair.first->getCanonicalType();
+      auto ReplacementTy = Pair.second;
+
+      if (!ReplacementTy->hasArchetype()) {
+        // Copy the entry.
+        copySubstitutionMapEntry(Ty, CalleeInterfaceToCallerArchetypeMap,
+                                 CalleeInterfaceToSpecializedArchetypeMap);
+        //assert(CalleeInterfaceToCallerArchetypeMap.getConformances(CanTy->getCanonicalType()).empty() &&
+        //       "Concrete type should not have any conformances");
+        continue;
+      }
+
+      // It is a substitution where the replacement contains an archetype.
+      auto SubstInterfaceReplacementTy =
+          ReplacementTy.subst(CallerArchetypeToSpecializedInterfaceMap)
+              ->getCanonicalType();
+      auto SubstReplacementTy = Builder.mapTypeIntoContext(
+          SM, SpecializedGenericEnv, SubstInterfaceReplacementTy);
+
+      CalleeInterfaceToSpecializedArchetypeMap.addSubstitution(
+          CanTy, SubstReplacementTy);
+
+      // Now remap conformances.
+      auto Conformances = CalleeInterfaceToCallerArchetypeMap.getConformances(CanTy);
+      SmallVector<ProtocolConformanceRef, 1> SubstConformances;
+      SmallVector<Substitution, 4> CallerArchetypeToSpecializedInterfaceSubs;
+      CallerGenericSig->getSubstitutions(
+          *SM, CallerInterfaceToSpecializedInterfaceMap,
+          CallerArchetypeToSpecializedInterfaceSubs);
+      for (auto C : Conformances) {
+        if (C.isAbstract()) {
+          SubstConformances.push_back(C);
+          continue;
+        }
+
+        auto *ConcreteC = C.getConcrete();
+        if (!ConcreteC)
+          continue;
+
+        auto SpecializedConformance =
+            dyn_cast<SpecializedProtocolConformance>(ConcreteC);
+
+        if (SpecializedConformance &&
+            SpecializedConformance->getRootNormalConformance()
+                ->getGenericSignature()) {
+          auto ConcreteSubs =
+              SubstReplacementTy->gatherAllSubstitutions(SM, nullptr, nullptr);
+          ConcreteC = Ctx.getSpecializedConformance(
+              SubstReplacementTy,
+              SpecializedConformance->getRootNormalConformance(),
+              ConcreteSubs);
+        }
```

Comment:
Sure, but it belongs on ProtocolConformanceRef. I'm *sure* other code is doing this, and we should move the abstractions where they'll get the most reuse.